index.d.ts

```diff
@@ -0,0 +1,18 @@
+import {Middleware, Dispatch} from "redux";
+
+
+export type ThunkAction<R, S, E> = (dispatch: Dispatch<S>, getState: () => S,
+                                    extraArgument: E) => R;
+
+declare module "redux" {
+  export interface Dispatch<S> {
+    <R, E>(asyncAction: ThunkAction<R, S, E>): R;
```

Comment:
Yes, but does it really help?

It can simply be:

```ts
type ThunkAction<R, S, E>  = (dispatch: Dispatch, getState: () => S, extraArgument: E) => R;
declare module 'redux' {
  export interface Dispatch {
    <R, S, E>(asyncAction: ThunkAction<R, S, E>): R;
  }
}
```
The point is the `Dispatch` does not gain anything by having the type `S`.